enterprise/frontend/src/embedding-sdk/cli/steps/setup-metabase-instance.ts

```diff
@@ -0,0 +1,207 @@
+import { exec as execCallback } from "child_process";
+
+import toggle from "inquirer-toggle";
+import fetch from "node-fetch";
+import ora from "ora";
+import { promisify } from "util";
+
+import type { CliOutput, CliStepMethod } from "embedding-sdk/cli/types/cli";
+import {
+  OUTPUT_STYLES,
+  printEmptyLines,
+  printInfo,
+} from "embedding-sdk/cli/utils/print";
+
+import { CONTAINER_NAME, SITE_NAME } from "../constants/config";
+import { EMBEDDING_DEMO_SETUP_TOKEN } from "../constants/env";
+import {
+  EMBEDDING_FAILED_MESSAGE,
+  INSTANCE_CONFIGURED_MESSAGE,
+} from "../constants/messages";
+import { retry } from "../utils/retry";
+
+const exec = promisify(execCallback);
+
+export const setupMetabaseInstance: CliStepMethod = async state => {
+  const spinner = ora();
+
+  const showError = (message: string): CliOutput => {
+    spinner.fail();
+    return [
+      {
+        type: "error",
+        message,
+      },
+      state,
+    ];
+  };
+
+  // If the instance we are configuring is not clean,
+  // therefore we cannot ensure the setup steps are performed.
+  const onInstanceConfigured = async (): Promise<CliOutput> => {
+    spinner.fail();
+    printEmptyLines();
+    printInfo(
+      "The instance is already configured. Do you want to delete the container and start over?",
+    );
+    const shouldRestartSetup = await toggle({
+      message: `${OUTPUT_STYLES.error("WARNING: This will delete all data.")}`,
+      default: false,
+    });
+    if (!shouldRestartSetup) {
+      return showError(INSTANCE_CONFIGURED_MESSAGE);
+    }
+    await exec(`docker rm -f ${CONTAINER_NAME}`);
+
+    return [
+      {
+        type: "success",
+        nextStep: "startLocalMetabaseContainer",
+      },
+      state,
+    ];
+  };
+
+  // The API is not immediately ready after the health check.
+  // We keep retrying until the request stops timing out.
+  try {
+    spinner.start("Creating an admin user (~2 mins)");
+
+    let res = await retry(
+      () =>
+        fetch(`${state.instanceUrl}/api/setup`, {
+          method: "POST",
+          body: JSON.stringify({
+            // Instead of fetching /api/session/properties, we can use the demo setup token.
+            token: EMBEDDING_DEMO_SETUP_TOKEN,
+            user: {
+              email: state.email,
+              password: state.password,
+              password_confirm: state.password,
+              last_name: "Embedding",
+              first_name: "Demo",
+              site_name: SITE_NAME,
+            },
+            prefs: {
+              site_name: SITE_NAME,
+              site_locale: "en",
+            },
+          }),
+          headers: { "content-type": "application/json" },
+          signal: AbortSignal.timeout(2500),
+        }),
+      { retries: 20, delay: 100 },
```

Comment:
Am I reading this correctly? This is waiting for 100ms before trying for 20 times that's about 2 seconds worth of wait time!?

It'd be more realistic if we're waiting like for 1 second before retrying. 100ms is too spammy IMO.

Reply:
changed it to 1000!

Reply:
@WiNloSt FYI, the reason for this was that even after the health checks has completed, the API timeouts for a long while before the API starts working. The small delay was to account for the period of time when the API wouldn't respond (can take 30 seconds to 2 minutes)

In my testing, the long delay between retries does slow down this sequence a little bit.